xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xpn.xwiki.internal.doc;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.ListIterator;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+import com.xpn.xwiki.internal.AbstractNotifyOnUpdateList;
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @param <E>
+ * @since 10.0RC1
+ */
+public class XWikiAttachmentList extends AbstractNotifyOnUpdateList<XWikiAttachment>
+{
+
+    private final Map<String, XWikiAttachment> map = new ConcurrentSkipListMap<String, XWikiAttachment>();
+
+    private final XWikiDocument document;
+
+    /**
+     * Initializes the map.
+     * 
+     * @since 10.0RC1
+     */
+    public XWikiAttachmentList(XWikiDocument document)
+    {
+        super(new ArrayList<XWikiAttachment>());
+        this.document = document;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        this.list = new ArrayList<XWikiAttachment>(map.values());
```

Comment:
Since you need to both:
* recreate the List
* update document metadata dirty flag

for each write operation you should probably have both in the same method called by all write operations.